### PR TITLE
TST Adds zero batch dims into ModuleInfo tests

### PR DESCRIPTION
**Pull request by thomasjpfan**
Related to https://github.com/pytorch/pytorch/issues/12013

cc @albanD @mruberry @jbschlosser @walterddr

### CODE REVIEW

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -300,6 +340,10 @@ def module_inputs_torch_nn_L1Loss(module_info, device, dtype, requires_grad, **k
                                                 make_input(shape=(2, 3, 4))),
                     reference_fn=lambda m, p, i, t: 1. / i.numel() * sum((a - b).abs().sum()
                                                                          for a, b in zip(i, t))),
+        ModuleInput(constructor_input=FunctionInput(),
+                    forward_input=FunctionInput(make_input(shape=(0, 3, 4)),
+                                                make_input(shape=(0, 3, 4))),
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:345]: Should the shapes here have 0s? I thought that was handled by the new decorator

File: torch/testing/_internal/common_modules.py [hunk 2]
```diff
@@ -158,6 +158,38 @@ def formatted_name(self):
...
+    return ModuleInput(
+        constructor_input=_zero_first_dim(module_input.constructor_input),
+        forward_input=_zero_first_dim(module_input.forward_input),
+        desc=module_input.desc + 'zero_batch_dim')
+
+def _module_inputs_with_zero_batch_dim(f):
+    # Creates ModuleInput by reducing the size of the 0th dimension to 0
+    # in constructor_input and forward_input of module_inputs[0]
+    # Make sure the tensors in module_inputs[0] semantically make sense with
+    # the 0th dimension set to size 0
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:184]: So this imposes a restriction on the `module_inputs_func` that the first `ModuleInput` needs to have a batch dim so this can zero it out?

I think it'd be a little less strange if the function applied directly to the first input. Maybe something like `duplicate_with_zero_batch_dim_input(module_input) -> [Tensor]` idk - wdyt?

Comment (thomasjpfan) [reply torch/testing/_internal/common_modules.py]: I updated the PR with a version where the function is applied directly to the first input and appended to `module_inputs` in `module_inputs_func`.

**Review by jbschlosser — APPROVED** [review]
LGTM!

**Comment by jbschlosser** [comment]
@thomasjpfan there's a failing UBSAN check regarding `NLLLoss` - is this coming from the generated zero batch dim input?

**Comment by thomasjpfan** [comment]
I updated the implementation to avoid dividing by zero and set `nan` explicitly.

Fun fact: `test_nll_loss_empty_tensor_reduction_mean` in `test_nn.py` is skipped when UBSAN is on.

**Comment by jbschlosser** [comment]
> I updated the implementation to avoid dividing by zero and set `nan` explicitly.
> 
> Fun fact: `test_nll_loss_empty_tensor_reduction_mean` in `test_nn.py` is skipped when UBSAN is on.

That is fun, and I guess we'd need sample-level tests to be able to skip at that granularity. When I get back from PTO, I'll look deeper at your PR exploring this

**Comment by thomasjpfan** [comment]
Closing because the work is continued at https://github.com/pytorch/pytorch/pull/68098
